parser/wasm_parse/mod.rs

```diff
@@ -73,81 +92,77 @@ impl<'a> Parse<'a> for wasmparser::ModuleReader<'a> {
             }
         }
 
-        for (idx, (size, section)) in sections.into_iter().enumerate() {
+        // Next, we parse the function and code sections together, so that we
+        // can collapse corresponding entries from the code and function
+        // sections into a single representative IR item.
+        match (function_section, code_section) {
+            (Some(function_section), Some(code_section)) => (function_section, code_section)
+                .parse_items(items, (imported_functions, &names, &sizes))?,
+            _ => panic!("function or code section is missing"),
+        };
+
+        for IndexedSection(idx, section) in sections.into_iter() {
             let start = items.size_added();
-            let name = match section.code {
+            let name = get_section_name(&section);
+            match section.code {
                 wasmparser::SectionCode::Custom { name, .. } => {
                     CustomSectionReader(name, section).parse_items(items, idx)?;
-                    format!("custom section '{}' headers", name)
                 }
                 wasmparser::SectionCode::Type => {
                     section.get_type_section_reader()?.parse_items(items, idx)?;
-                    "type section headers".to_string()
                 }
                 wasmparser::SectionCode::Import => {
                     section
                         .get_import_section_reader()?
                         .parse_items(items, idx)?;
-                    "import section headers".to_string()
                 }
                 wasmparser::SectionCode::Function => {
-                    section
-                        .get_function_section_reader()?
-                        .parse_items(items, idx)?;
-                    "function section headers".to_string()
+                    panic!("unexpected function section");
```

Comment:
tangentially, not _entirely_ sure what to do right here. We shouldn't ever hit this branch.

Reply:
If this branch is actually unreachable, as opposed to something that can happen but is an error, then we should use the `unreachable!()` macro. I believe that is the case here, right?